File: scripts/utils/paths.R
#!/usr/bin/env Rscript
# Path Utilities
# Centralized path management for the project

# Get project root directory
# This function searches upward from current directory to find project root
get_project_root <- function() {
  current_dir <- getwd()
  max_depth <- 10
  depth <- 0

  while (depth < max_depth) {
    # Check for project root indicators
    has_readme <- file.exists(file.path(current_dir, "README.md"))
    has_bin <- file.exists(file.path(current_dir, "bin"))
    has_scripts <- file.exists(file.path(current_dir, "scripts"))
    has_data <- file.exists(file.path(current_dir, "data"))

    if (has_readme && has_bin && has_scripts && has_data) {
      return(normalizePath(current_dir))
    }

    parent_dir <- dirname(current_dir)
    if (parent_dir == current_dir) {
      break  # Reached filesystem root
    }
    current_dir <- parent_dir
    depth <- depth + 1
  }

  # Fallback: return current directory
  return(normalizePath(getwd()))
}

# Set project root
PROJECT_ROOT <- get_project_root()

# Directory paths
DATA_DIR <- file.path(PROJECT_ROOT, "data")
RESEARCH_DIR <- file.path(PROJECT_ROOT, "research")
EVIDENCE_DIR <- file.path(PROJECT_ROOT, "evidence")
FILINGS_DIR <- file.path(PROJECT_ROOT, "filings")
SCRIPTS_DIR <- file.path(PROJECT_ROOT, "scripts")
BIN_DIR <- file.path(PROJECT_ROOT, "bin")
CONFIG_DIR <- file.path(PROJECT_ROOT, "config")
OUTPUTS_DIR <- file.path(PROJECT_ROOT, "outputs")

# Data subdirectories
DATA_SOURCE_DIR <- file.path(DATA_DIR, "source")
DATA_RAW_DIR <- file.path(DATA_DIR, "raw")
DATA_CLEANED_DIR <- file.path(DATA_DIR, "cleaned")
DATA_ANALYSIS_DIR <- file.path(DATA_DIR, "analysis")
DATA_SCRAPED_DIR <- file.path(DATA_DIR, "scraped")
DATA_VECTORS_DIR <- file.path(DATA_DIR, "vectors")

# Research subdirectories
RESEARCH_CONNECTIONS_DIR <- file.path(RESEARCH_DIR, "connections")
RESEARCH_VIOLATIONS_DIR <- file.path(RESEARCH_DIR, "violations")
RESEARCH_ANOMALIES_DIR <- file.path(RESEARCH_DIR, "anomalies")
RESEARCH_EVIDENCE_DIR <- file.path(RESEARCH_DIR, "evidence")
RESEARCH_VERIFICATION_DIR <- file.path(RESEARCH_DIR, "verification")
RESEARCH_TIMELINES_DIR <- file.path(RESEARCH_DIR, "timelines")
RESEARCH_SUMMARIES_DIR <- file.path(RESEARCH_DIR, "summaries")
RESEARCH_SEARCH_RESULTS_DIR <- file.path(RESEARCH_DIR, "search_results")

# Config files
STATE_REGISTRY_FILE <- file.path(CONFIG_DIR, "state_dpor_registry.csv")

# Helper function to source scripts
source_script <- function(script_path) {
  full_path <- file.path(SCRIPTS_DIR, script_path)
  if (file.exists(full_path)) {
    source(full_path)
  } else {
    stop(paste("Script not found:", full_path))
  }
}

# Helper function to source bin scripts
source_bin <- function(script_name) {
  full_path <- file.path(BIN_DIR, script_name)
  if (file.exists(full_path)) {
    source(full_path)
  } else {
    stop(paste("Bin script not found:", full_path))
  }
}

File: scripts/utils/paths.py
#!/usr/bin/env python3
"""
Path Utilities for Python Scripts

Provides consistent path management across the project.
Uses Python 3.14 features: match expressions, modern type hints.
"""

from pathlib import Path

# Find project root using match expression (Python 3.14+)
def find_project_root() -> Path:
    """Find the project root directory."""
    current = Path(__file__).resolve()

    # Check immediate parents first
    for parent in [current.parent.parent.parent, current.parent.parent]:
        if (parent / "README.md").exists() and (parent / "bin").exists():
            return parent

    # Fallback: search from current directory
    current_dir = Path.cwd()
    for _ in range(10):
        if (current_dir / "README.md").exists() and (current_dir / "bin").exists():
            return current_dir
        if current_dir == current_dir.parent:
            break
        current_dir = current_dir.parent

    return Path.cwd()

# Set project root
PROJECT_ROOT = find_project_root()

# Directory paths using Path operations
BIN_DIR = PROJECT_ROOT / "bin"
SCRIPTS_DIR = PROJECT_ROOT / "scripts"
DATA_DIR = PROJECT_ROOT / "data"
RESEARCH_DIR = PROJECT_ROOT / "research"
EVIDENCE_DIR = PROJECT_ROOT / "evidence"
FILINGS_DIR = PROJECT_ROOT / "filings"
OUTPUTS_DIR = PROJECT_ROOT / "outputs"
DOCS_DIR = PROJECT_ROOT / "docs"

# Data subdirectories
DATA_SOURCE_DIR = DATA_DIR / "source"
DATA_RAW_DIR = DATA_DIR / "raw"
DATA_CLEANED_DIR = DATA_DIR / "cleaned"
DATA_SCRAPED_DIR = DATA_DIR / "scraped"
DATA_VECTORS_DIR = DATA_DIR / "vectors"
DATA_PROCESSED_DIR = DATA_DIR / "processed"
DATA_PROCESSED_DIR = DATA_DIR / "processed"

# Research subdirectories
RESEARCH_CONNECTIONS_DIR = RESEARCH_DIR / "connections"
RESEARCH_VIOLATIONS_DIR = RESEARCH_DIR / "violations"
RESEARCH_ANOMALIES_DIR = RESEARCH_DIR / "anomalies"
RESEARCH_EVIDENCE_DIR = RESEARCH_DIR / "evidence"
RESEARCH_VERIFICATION_DIR = RESEARCH_DIR / "verification"
RESEARCH_TIMELINES_DIR = RESEARCH_DIR / "timelines"
RESEARCH_SUMMARIES_DIR = RESEARCH_DIR / "summaries"
RESEARCH_SEARCH_RESULTS_DIR = RESEARCH_DIR / "search_results"

# Ensure directories exist using efficient pattern
def ensure_directories() -> None:
    """Ensure all necessary directories exist."""
    directories = [
        DATA_DIR, DATA_SOURCE_DIR, DATA_RAW_DIR, DATA_CLEANED_DIR,
        DATA_SCRAPED_DIR, DATA_VECTORS_DIR, DATA_PROCESSED_DIR,
        RESEARCH_DIR, RESEARCH_CONNECTIONS_DIR, RESEARCH_VIOLATIONS_DIR,
        RESEARCH_ANOMALIES_DIR, RESEARCH_EVIDENCE_DIR, RESEARCH_VERIFICATION_DIR,
        RESEARCH_TIMELINES_DIR, RESEARCH_SUMMARIES_DIR, RESEARCH_SEARCH_RESULTS_DIR,
        OUTPUTS_DIR, FILINGS_DIR
    ]

    # Use list comprehension for efficiency
    [directory.mkdir(parents=True, exist_ok=True) for directory in directories]

# Initialize directories on import
ensure_directories()
